### DISCUSSION TITLE
It it possible to provide the rc4 decodekey from outside the code?

### DISCUSSION THREAD

**Discussion by Kezzel** [javascript-obfuscator/javascript-obfuscator#1254] [Q&A]
It seems relatively easy to deobfuscate code using online tools. The result is still gibberish, however the used strings can be read. This shows a little too much of the inner workings of our code.

The deopfuscation process seems to be able to find the decodekey for the rc4 directly in the code. So my question is, if it is possible to provide this key as an option when obfuscating, and then provide it, fx. as an variable on the script tag used to run the obfuscated script and then fetch this key using the currentScript global.

If the script is removed after load, it becomes extremely difficult to even see that such key is provided, and it would make the deobfuscation much harder.

What do you think?

Cheers

**Comment by BlackYuzia** [comment]
Suppose, that nothing hard to find external key

**Reply by Kezzel** [reply]
The key is provided by chrome extension code, where the code is loaded through injectScript. 

It will never be impossible to deobfuscate, I guess that is the nature of js, but if it is not possible to simply use an online tool that recognizes that it has been obfuscated using the javascript-obuscator tool and deobfuscates it right away.

Hiding the key elsewhere will make it much harder, and require real effort.

**Reply by BlackYuzia** [reply]
Don't sure on 100%. But, suppose external key is like:
`const key =  await fetchKey()`?
Or what you mean instead?

**Reply by Kezzel** [reply]
Well, I would like to provide a key string to be used when obfuscating (as an option) and a function that retrieves the key from within the script, again as an option. Fx.:
`{
stringArrayEncodingRc4Key: 'some-key',
//If the key fx. is stored as the id value of the script inserted
stringArrayEncodingRc4KeyFunction: 'currentScript.attributes.id.value'
}`


It could also be a global variable, or fetched async from somewhere etc.:-)

**Reply by BlackYuzia** [reply]
Same as right now?

**Reply by Kezzel** [reply]
No a I mean in the global scope, fx. set on global/window object. The point is that the key is NOT part of the script itself, and therefore the script can not on its own be deobfuscated, but requires at least the detective task to figure out from where / how the key is obtained.
